openedx/core/djangoapps/programs/utils.py

```diff
@@ -460,57 +461,86 @@ def _attach_course_run_upgrade_url(self, run_mode):
+        """
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
+        )
+        course_runs_with_enrollments = set(unicode(enrollment.course_id) for enrollment in enrollments)
+        courses_without_enrollments = []
+        for course in courses:
+            if len([run for run in course['course_runs'] if unicode(run['key']) in course_runs_with_enrollments]) == 0:
+                courses_without_enrollments.append(course)
+
+        return courses_without_enrollments
 
     def _collect_one_click_purchase_eligibility_data(self):
         """
         Extend the program data with data about learner's eligibility for one click purchase,
         discount data of the program and SKUs of seats that should be added to basket.
         """
-        applicable_seat_types = self.data['applicable_seat_types']
+        applicable_seat_types = set(seat for seat in self.data['applicable_seat_types'] if seat != 'credit')
         is_learner_eligible_for_one_click_purchase = self.data['is_program_eligible_for_one_click_purchase']
         skus = []
         bundle_variant = 'full'
-        if is_learner_eligible_for_one_click_purchase:
-            for course in self.data['courses']:
-                add_course_sku = True
-                course_runs = course.get('course_runs', [])
-                published_course_runs = filter(lambda run: run['status'] == 'published', course_runs)
-
-                if len(published_course_runs) == 1:
-                    for course_run in course_runs:
-                        is_paid_seat = self._check_enrollment_for_user(course_run)
 
-                        if is_paid_seat:
-                            add_course_sku = False
-                            break
-
-                    if add_course_sku:
+        if is_learner_eligible_for_one_click_purchase:
+            courses = self._filter_out_courses_with_enrollments(self.data['courses'])
+            courses = self._filter_out_courses_with_entitlements(courses)
+
+            if len(courses) < len(self.data['courses']):
+                bundle_variant = 'partial'
+
+            for course in courses:
+                entitlements = []
+                for entitlement in course['entitlements']:
+                    if entitlement['mode'].lower() in applicable_seat_types:
+                        entitlements.append(entitlement)
+                if entitlements:
+                    skus.append(entitlements[0]['sku'])
+                else:
+                    course_runs = course.get('course_runs', [])
+                    published_course_runs = [run for run in course_runs if run['status'] == 'published']
+                    if len(published_course_runs) == 1:
                         for seat in published_course_runs[0]['seats']:
                             if seat['type'] in applicable_seat_types and seat['sku']:
```

Comment:
Do we need to call `.lower()` on `seat[`type`]`

Reply:
No because the seat class has an enumerated list of possible values for the type field, all of which are lowercased.

Reply:
@amangano-edx In LMS it is assumed throughout the code base that the mode is correct.  We should avoid adding lower() into it as it could mask errors in the data.